File: samples/csharp_dotnetcore/55.teams-link-unfurling/Bots/LinkUnfurlingBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Teams;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;

namespace Microsoft.BotBuilderSamples.Bots
{
    public class LinkUnfurlingBot : TeamsActivityHandler
    {
        protected override Task<MessagingExtensionResponse> OnTeamsAppBasedLinkQueryAsync(ITurnContext<IInvokeActivity> turnContext, AppBasedLinkQuery query, CancellationToken cancellationToken)
        {
            var heroCard = new ThumbnailCard
            {
                Title = "Thumbnail Card",
                Text = query.Url,
                Images = new List<CardImage> { new CardImage("https://raw.githubusercontent.com/microsoft/botframework-sdk/master/icon.png") },
            };

            var attachments = new MessagingExtensionAttachment(HeroCard.ContentType, null, heroCard);
            var result = new MessagingExtensionResult("list", "result", new[] { attachments });

            return Task.FromResult(new MessagingExtensionResponse(result));
        }

        protected override Task<MessagingExtensionResponse> OnTeamsMessagingExtensionQueryAsync(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionQuery query, CancellationToken cancellationToken)
        {
            //Note: The Teams manifest.json for this sample also includes a Search Query, in order to enable installing from App Studio.

            switch (query.CommandId)
            {
                // These commandIds are defined in the Teams App Manifest.
                case "searchQuery":
                    var card = new HeroCard
                    {
                        Title = "This is a Link Unfurling Sample",
                        Subtitle = "It will unfurl links from *.BotFramework.com",
                        Text = "This sample demonstrates how to handle link unfurling in Teams.  Please review the readme for more information.",
                    };

                    return Task.FromResult(new MessagingExtensionResponse
                    {
                        ComposeExtension = new MessagingExtensionResult
                        {
                            AttachmentLayout = "list",
                            Type = "result",
                            Attachments = new List<MessagingExtensionAttachment>
                            {
                                new MessagingExtensionAttachment
                                {
                                    Content = card,
                                    ContentType = HeroCard.ContentType,
                                    Preview = card.ToAttachment(),
                                },
                            },
                        },
                    });

                default:
                    throw new NotImplementedException($"Invalid CommandId: {query.CommandId}");
            }
        }
    }
}

File: archive/samples/python/55.teams-link-unfurling/bots/link_unfurling_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core import TurnContext, CardFactory
from botbuilder.core.card_factory import ContentTypes
from botbuilder.core.teams import TeamsActivityHandler
from botbuilder.schema import ThumbnailCard, CardImage, HeroCard
from botbuilder.schema.teams import (
    AppBasedLinkQuery,
    MessagingExtensionQuery,
    MessagingExtensionAttachment,
    MessagingExtensionResult,
    MessagingExtensionResponse,
)


class LinkUnfurlingBot(TeamsActivityHandler):
    async def on_teams_app_based_link_query(
        self, turn_context: TurnContext, query: AppBasedLinkQuery
    ):
        hero_card = ThumbnailCard(
            title="Thumbnail Card",
            text=query.url,
            images=[
                CardImage(
                    url="https://raw.githubusercontent.com/microsoft/botframework-sdk/master/icon.png"
                )
            ],
        )

        attachments = MessagingExtensionAttachment(
            content_type=ContentTypes.hero_card, content=hero_card
        )

        result = MessagingExtensionResult(
            attachment_layout="list", type="result", attachments=[attachments]
        )

        return MessagingExtensionResponse(compose_extension=result)

    async def on_teams_messaging_extension_query(
        self, turn_context: TurnContext, query: MessagingExtensionQuery
    ):
        # These commandIds are defined in the Teams App Manifest.
        if not query.command_id == "searchQuery":
            raise NotImplementedError(f"Invalid CommandId: {query.command_id}")

        card = HeroCard(
            title="This is a Link Unfurling Sample",
            subtitle="It will unfurl links from *.BotFramework.com",
            text="This sample demonstrates how to handle link unfurling in Teams.  Please review the readme for more "
            "information. ",
        )

        return MessagingExtensionResponse(
            compose_extension=MessagingExtensionResult(
                attachment_layout="list",
                type="result",
                attachments=[
                    MessagingExtensionAttachment(
                        content=card,
                        content_type=ContentTypes.hero_card,
                        preview=CardFactory.hero_card(card),
                    )
                ],
            )
        )
